### ISSUE TITLE
Enable fade of Auto Bed Leveling

### ISSUE THREAD

**Issue by DzzD** [bambulab/BambuStudio#1602]
Marlin enable to use the special command `M420 S1 Z10` to fade the ABL on a certain height, would be very nice if this was possible on Bambulab Printer to use this gcode or another to use fading on ABL.

For example :
`M420 S1 Z10`

Would mean that ABL is used 
- 100% when z = 0mm
- 50% when z = 5mm
- 0% when z = 10mm (at this point ABL may be disabled too)

Note that this fade can be done within the firmware using a special gcode or in the studio by modifying the generated GCODE when slicing.

**Comment by Majestic7979** [comment]
Can you explain why this is even necessary? With a solid example. What hurts the print if this isn't provided?

**Comment by DzzD** [comment]
> Can you explain why this is even necessary? With a solid example. What hurts the print if this isn't provided?

heuuu... strange question, but yes I can, it's a well know option in slicer and/or firmware, it make the part perfectly flat at a given height, and then the printer can act exacly like if the printer bed was perfectly flat, but that's only one advantage.

For example on a completly warped bed you can print a raft that is 5mm thick and with a z fade of 5mm, it will be perfeclty flat on top even if the bed is completly warp, but there are tons of other usage, just do some search, it is not mandatory, but like many options like bed levelling, this is a very basic feature wich all good slicer or firmware should have...

Another exemple is that when you have cut two parts and want then to clip perfeclty on their top surface by the plug you create/added, etc, etc...

**Comment by Majestic7979** [comment]
Ok thanks. It's not a strange question. I wasn't aware of what this option was for. I saw it on Klipper when I was building my config file but never really occurred to me that it would affect stuff. So basically you're telling the printer to fill the gaps at the top from what I understood of your explanation so that it ends up flat. Is that because you assume that the part won't be flat unless there is fading? Or is this actually documented in firmware?

**Comment by DzzD** [comment]
You ask the firmware to reduce ABL correction each new layer until there is no more need of ABL as layer are totaly flat, it is usally set from 4mm to 10mm : 

https://3dprinting.stackexchange.com/questions/20830/what-is-the-fade-height-option-in-marlin

**Comment by DzzD** [comment]
They sure know about this feature as it is a common feature in FDM printing... but anyway it is not a self service, I have posted this issue because I think it will be a good for the Studio, not for myself, I can live without

**Comment by DzzD** [comment]
Anyway, I dont really understand what you are trying to teach me, and why you try to explain me how this and the world should works, I created my first public opensource project back in 2005. They can gently ignore my isssue if they find it useless for the community.

**Comment by DzzD** [comment]
Did you tryied "Precise Z Height" ?

![Image](https://github.com/user-attachments/assets/6b2da3e4-e726-4ada-9d6a-dbf7afffc008)

https://wiki.bambulab.com/en/software/bambu-studio/layer-height#precise-z-height

**Comment by ValCT** [comment]
Yes, it's always ON, but it has an impact on the very last layers, I would like things to change for the whole print exept the few first layers

**Comment by Majestic7979** [comment]
> Anyway, I dont really understand what you are trying to teach me, and why you try to explain me how this and the world should works, I created my first public opensource project back in 2005. They can gently ignore my isssue if they find it useless for the community.

I just noticed a notification for this... I suggest you to relax, ok? I was not asking questions to attack if that's what you're thinking. You are aware that development time is limited, developers are resources, think of them like a sack of flour. You can only use so much flour to make pancakes, if you also want to make bread then you need to be rational and explain why making more pancakes is a good idea, or else there won't be able more flour to make bread.

I was thinking like a developer or product owner. It doesn't matter if they know or don't know. It matters WHY you are here asking for this to be implement. It will cost development time.

So my intention was to steer you to make your case.

But you didn't help yourself at all by saying "they can ignore the request". When you use that kind of language then that is exactly what they're gonna do... In the end it ended being a waste of your time writing the request. It's your time but they also wasted their time reading it only to learn several messages later that you don't really need it.

Please understand that my messages are not related to you and it's not personal. They're about gathering info and trying to show BL why your request is worth their consideration.

Peter

**Comment by DzzD** [comment]
> > Anyway, I dont really understand what you are trying to teach me, and why you try to explain me how this and the world should works, I created my first public opensource project back in 2005. They can gently ignore my isssue if they find it useless for the community.
> 
> I just noticed a notification for this... I suggest you to relax, ok? I was not asking questions to attack if that's what you're thinking. You are aware that development time is limited, developers are resources, think of them like a sack of flour. You can only use so much flour to make pancakes, if you also want to make bread then you need to be rational and explain why making more pancakes is a good idea, or else there won't be able more flour to make bread.
> 
> I was thinking like a developer or product owner. It doesn't matter if they know or don't know. It matters WHY you are here asking for this to be implement. It will cost development time.
> 
> So my intention was to steer you to make your case.
> 
> But you didn't help yourself at all by saying "they can ignore the request". When you use that kind of language then that is exactly what they're gonna do... In the end it ended being a waste of your time writing the request. It's your time but they also wasted their time reading it only to learn several messages later that you don't really need it.
> 
> Please understand that my messages are not related to you and it's not personal. They're about gathering info and trying to show BL why your request is worth their consideration.
> 
> Peter

ho ! that's old, it seems that I was in a bad mood and nervous when replied :) :)

**Comment by ValCT** [comment]
I may have a workaround for those who, like me, need to print technical parts with really accurate dimensions, what seem obviously compromise by the lack of an ABL fade option.

I saw in the bed tramming gcode, provided by Bambu Lab, that the M29 command could be used to disable ABL compensation (so, that wasn't M420) :

`M29.2 S0 ; turn off pos comp` (did someone ever see such a command, with a .2 after the M## ?)

The question is, how to use it :

- In the start gcode, but to avoid bed scratches, one have to make sure that the bed defects are smaller than the thickness of the first layer height (mine are bigger than 0.4mm), and to ensure adhesion, use an excessive flow rate (just like we used to proceed with the old printers, anyway...).
- Use a raft and manually insert this line in the file after the first layer (could it be done automatically within the slicer ?)

Any opinions ?

**Comment by DzzD** [comment]
> `M29.2 S0 ; turn off pos comp` (did someone ever see such a command, with a .2 after the M## ?)

This is sometime used (like in GRBL G28.1/G92.1/etc... for CNC), it is kind of sub-command

>Any opinions ?

Disabling ABL should work on non-warpped bed of for any small area part, I would just suggest to print at the center of the bed where the Z homing is done

**Comment by ValCT** [comment]
Thanks ! That is exactly the kind of things I print, so, let's try out this
way.

Le dim. 16 févr. 2025 à 15:12, Bruno Augier ***@***.***> a
écrit :
